lib/ecto/multi.ex

```diff
@@ -364,37 +368,56 @@ defmodule Ecto.Multi do
+      operations
+      |> Enum.reduce({%{}, names}, &apply_operation(&1, repo, wrap, return, &2))
+      |> elem(0)
     end)
   end
 
-  defp apply_operations({:error, error}, _repo, _wrap, _return) do
+  defp apply_operations({:error, error}, _names, _repo, _wrap, _return) do
     {:error, error}
   end
 
-  defp apply_operation({name, operation}, repo, return, acc) do
-    case apply_operation(operation, acc, repo) do
+  defp apply_operation({name, operation}, repo, wrap, return, {acc, names}) do
+    case apply_operation(operation, acc, {wrap, return}, repo) do
+      {:merge, {:ok, value}} ->
+        merge_results(acc, value, names)
+      {:merge, {:error, {name, value, nested_acc}}} ->
+        {acc, _names} = merge_results(acc, nested_acc, names)
+        return.({name, value, acc})
       {:ok, value} ->
-        Map.put(acc, name, value)
+        {Map.put(acc, name, value), names}
       {:error, value} ->
         return.({name, value, acc})
     end
   end
 
-  defp apply_operation({:changeset, changeset, opts}, _acc, repo),
+  defp apply_operation({:changeset, changeset, opts}, _acc, _apply_args, repo),
     do: apply(repo, changeset.action, [changeset, opts])
-  defp apply_operation({:run, {mod, fun, args}}, acc, _repo),
-    do: apply(mod, fun, [acc | args])
-  defp apply_operation({:run, run}, acc, _repo),
-    do: apply(run, [acc])
-  defp apply_operation({:changeset_fun, action, run, opts}, acc, repo),
-    do: apply_operation({:changeset, put_action(run.(acc), action), opts}, acc, repo)
-  defp apply_operation({:insert_all, source, entries, opts}, _acc, repo),
+  defp apply_operation({:run, run}, acc, _apply_args, _repo),
+    do: apply_fun(run, acc)
+  defp apply_operation({:insert_all, source, entries, opts}, _acc, _apply_args, repo),
     do: {:ok, repo.insert_all(source, entries, opts)}
-  defp apply_operation({:update_all, query, updates, opts}, _acc, repo),
+  defp apply_operation({:update_all, query, updates, opts}, _acc, _apply_args, repo),
     do: {:ok, repo.update_all(query, updates, opts)}
-  defp apply_operation({:delete_all, query, opts}, _acc, repo),
+  defp apply_operation({:delete_all, query, opts}, _acc, _apply_args, repo),
     do: {:ok, repo.delete_all(query, opts)}
+  defp apply_operation({:merge, merge}, acc, {wrap, return}, repo),
+    do: {:merge, __apply__(apply_fun(merge, acc), repo, wrap, return)}
+
+  defp apply_fun({mod, fun, args}, acc), do: apply(mod, fun, [acc | args])
+  defp apply_fun(fun, acc),              do: apply(fun, [acc])
+
+  defp merge_results(changes, new_changes, names) do
+    new_names = new_changes |> Map.keys |> MapSet.new
+    case MapSet.intersection(names, new_names) |> MapSet.to_list do
+      [] ->
+        {Map.merge(changes, new_changes), MapSet.union(names, new_names)}
+      common ->
+        raise "merging results of multi, found operations: #{inspect common} " <>
+          "already present in the enclosing multi"
```

Comment:
"cannot merge multi, the following operations were found in both Ecto.Multi: #{inspect common}"

?